### PR TITLE
[CA-1180] [CA-1179] Custom IAM roles scripts

### PR DESCRIPTION
**Pull request by andy7i**
https://broadworkbench.atlassian.net/browse/CA-1180
https://broadworkbench.atlassian.net/browse/CA-1179

Custom IAM roles to be set up in each environment.

Removed roles/billing.projectManager to prevent users from changing the billing account manually (since Terra will handle this for users moving forward)

I made a spreadsheet that helped me track the permissions (it's a bit messy so I'd be happy to walk through it with folks): https://docs.google.com/spreadsheets/d/1n4BnfYIbEFSybKvh8WGbgbGNhYub8SdZexya8Pf5p4k/edit?pli=1#gid=610210961

The PRs related to this set of changes are:
https://github.com/broadinstitute/rawls/pull/1369
https://github.com/broadinstitute/terraform-terra/pull/144
https://github.com/broadinstitute/firecloud-develop/pull/2445
https://github.com/DataBiosphere/terra-resource-buffer/pull/120

### CODE REVIEW

File: scripts/project-setup-scripts/google-project-owner.yaml [hunk 1]
```diff
@@ -0,0 +1,1686 @@
...
+- workflows.operations.get
+- workflows.operations.list
+- workflows.workflows.get
+- workflows.workflows.getIamPolicy
+- workflows.workflows.list
+# </viewer> End of permissions included as part of roles/viewer
+# Add permissions for the other roles we need:
+##roles/billing.projectManager (updated 2/10/2021)
+- resourcemanager.projects.createBillingAssignment
+- resourcemanager.projects.deleteBillingAssignment
```
Comment (andy7i) [inline scripts/project-setup-scripts/google-project-owner.yaml:1671]: I think we want to remove these based on the fact that users will no longer be setting billing accounts on projects and Terra will handle that for users

File: scripts/project-setup-scripts/add-folder-roles-to-service-accounts.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# set up folder access for Rawls and Cromwell SAs as editor roles
+# must have resourcemanager.projects.setIamPolicy permissions
```
Comment (gpolumbo-broad) [inline scripts/project-setup-scripts/add-folder-roles-to-service-accounts.sh:2]: Do you mean that the person running this script needs this permission?  Maybe reword this as something like:

```suggestion
# In order to successfully run this script, you must first `gcloud auth login` as a gcloud user with resourcemanager.projects.setIamPolicy permissions
```

File: scripts/project-setup-scripts/add-folder-roles-to-service-accounts.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# env: staging
+# folder: test.firecloud.org/staging
+gcloud resource-manager folders add-iam-policy-binding 362889920837 --member=serviceAccount:44976466195-ic7tfqh2s0pn1v46nlc46c0gau6e8n60@developer.gserviceaccount.com --role=roles/editor
+gcloud resource-manager folders add-iam-policy-binding 362889920837 --member=serviceAccount:44976466195-nv2hrk6imhfccg4lftjd8p683hdampa1@developer.gserviceaccount.com --role=roles/editor
+
+# todo: to be completed as part of https://broadworkbench.atlassian.net/browse/CA-1194
+# env: prod
+# folder: firecloud.org/prod
+#gcloud resource-manager folders add-iam-policy-binding FOLDER --member=serviceAccount:rawls-prod@broad-dsde-prod.iam.gserviceaccount.com --role=roles/editor
+#gcloud resource-manager folders add-iam-policy-binding FOLDER --member=serviceAccount:cromwell-prod@broad-dsde-prod.iam.gserviceaccount.com --role=roles/editor
```
Comment (gpolumbo-broad) [inline scripts/project-setup-scripts/add-folder-roles-to-service-accounts.sh:36]: Feedback for the script as a whole:

I'm concerned that we're doing all the environments all at once.  First, it means that we cannot be targeted in our testing.  If I'm trying to run this on dev just to understand it or debug or whatever, I need to come in here and comment out the environments I do not want updated.   

Second, whomever I'm `gcloud auth` logged in as hopefully (and intentionally) should not have access to make all the specified changes.  For example, to make changes in prod, I should be required to use my `firecloud.org` account, which (rightfully so) should not have access to any other environment.  Same thing if I'm making changes to non-prod environments.  This means whenever I run this script, I'm expecting it to throw at least some errors, which could lead to _actual_ errors being overlooked or ignored.  

First thing off the top of my head to solve for the above is maybe a JSON or some other "config" file that is organized by environment and lists the SAs and Folder for each env.  Then when you run this script, what it does is read the appropriate config.  Or maybe these SA's and Folders are already stored in vault or something and can be read from there?  That might be even more "correct" in that Vault seems to be the source of truth for this sort of information.  Like if the Rawls SA changes for some reason, that change is going to be made in Vault, so if this script reads from the same place, we just rerun this without needing to make any modifications.

File: scripts/project-setup-scripts/create-custom-iam-roles-for-projects.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+gcloud iam roles create google-project-viewer --organization=400176686919 --file=google-project-viewer.yaml
+gcloud iam roles create google-project-owner --organization=400176686919 --file=google-project-owner.yaml
+
+# org: quality.firecloud.org
+gcloud iam roles create google-project-viewer --organization=206744735509 --file=google-project-viewer.yaml
+gcloud iam roles create google-project-owner --organization=206744735509 --file=google-project-owner.yaml
+
+# org: firecloud.org
+gcloud iam roles create google-project-viewer --organization=386193000800 --file=google-project-viewer.yaml
+gcloud iam roles create google-project-owner --organization=386193000800 --file=google-project-owner.yaml
```
Comment (gpolumbo-broad) [inline scripts/project-setup-scripts/create-custom-iam-roles-for-projects.sh:15]: Same feedback here, my preference would be that the script take an `organization` parameter

File: scripts/project-setup-scripts/google-project-owner.yaml [hunk 5]
```diff
@@ -0,0 +1,1683 @@
...
+- logging.logEntries.create
+##roles/monitoring.metricWriter (updated 2/10/2021)
+- monitoring.metricDescriptors.create
+- monitoring.metricDescriptors.get
+- monitoring.metricDescriptors.list
+- monitoring.monitoredResourceDescriptors.get
+- monitoring.monitoredResourceDescriptors.list
+- monitoring.timeSeries.create
+##RequesterPays custom role (updated 2/10/2021)
+- serviceusage.services.use
```
Comment (gpolumbo-broad) [inline scripts/project-setup-scripts/google-project-owner.yaml:1683]: There's a LOT getting set here.  Yes it's just read access for a lot of it, but still.  I think Bernick had made a request at one point to review permissions on things to make sure we're truly granting only permissions we need.  In other words, start with 0 permissions and only grant what's needed, as opposed to grant EVERYTHING and then remove permissions to things you don't want people getting at.  It seems like we could go through and bulk remove permissions for things like, `ml`, `gameservices`, `aiplatform`, etc.  If we need to add things in the future, it will be easy to do now that we have custom roles.

This seems like something InfoSec will care about too.  I feel like it should be pretty easy for us to chop hundreds of permissions away that we know we don't need these days.

File: scripts/project-setup-scripts/google-project-viewer.yaml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+- monitoring.monitoredResourceDescriptors.list
+- monitoring.timeSeries.create
+##RequesterPays custom role (updated 2/10/2021)
+- serviceusage.services.use
+##roles/bigquery.readSessionUser (updated 2/25/2021)
+- bigquery.readsessions.create
+- bigquery.readsessions.getData
+- bigquery.readsessions.update
+- resourcemanager.projects.get
+- resourcemanager.projects.list
```
Comment (gpolumbo-broad) [inline scripts/project-setup-scripts/google-project-viewer.yaml:26]: There are some duplicates in here.  Might not matter functionally, but for readability, I would prefer to not have them.

**Review by gpolumbo-broad — CHANGES_REQUESTED** [review]

File: scripts/project-setup-scripts/google-project-owner.yaml [hunk 7]
```diff
@@ -0,0 +1,1683 @@
+title: "google_project_owner"
```
Comment (dvoet) [inline scripts/project-setup-scripts/google-project-owner.yaml:1]: I would call this `terra_project_owner`

File: scripts/project-setup-scripts/add-folder-roles-to-service-accounts.sh (hunk 3 above)
Comment (dvoet) [reply scripts/project-setup-scripts/add-folder-roles-to-service-accounts.sh:36]: grab the SAs from vault

File: scripts/project-setup-scripts/google-project-viewer.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+title: "google_project_viewer"
```
Comment (dvoet) [inline scripts/project-setup-scripts/google-project-viewer.yaml:1]: `terra_workspace_writer`

File: scripts/project-setup-scripts/terra-billing-project-owner.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+title: "terra_billing_project_owner"
+description: "permissions for billing project owners"
+stage: "GA"
+includedPermissions:
```
Comment (andy7i) [inline scripts/project-setup-scripts/terra-billing-project-owner.yaml:4]: I've pruned this list in several passes, getting progressively more aggressive each pass. See commit history for details

File: scripts/project-setup-scripts/google-project-viewer.yaml (hunk 8 above)
Comment (andy7i) [reply scripts/project-setup-scripts/google-project-viewer.yaml]: after looking at this again, I'm wondering if "terra_workspace_can_compute" might be most accurate. Not all writers have can-compute

File: scripts/project-setup-scripts/google-project-owner.yaml (hunk 5 above)
Comment (andy7i) [reply scripts/project-setup-scripts/google-project-owner.yaml]: I've pruned this list in several passes, getting progressively more aggressive each pass. See commit history for details!

File: scripts/project-setup-scripts/google-project-viewer.yaml (hunk 8 above)
Comment (andy7i) [reply scripts/project-setup-scripts/google-project-viewer.yaml]: I went ahead and made the change but can revert it if we decide to

File: scripts/project-setup-scripts/terra-billing-project-owner.yaml (hunk 9 above)
Comment (dvoet) [reply scripts/project-setup-scripts/terra-billing-project-owner.yaml:4]: I am inclined to start with the yolo level and start with some communication that we are tightening security and users can contact us if we are too aggressive... adding permissions back in is easy

**Review by dvoet — APPROVED** [review]

File: scripts/project-setup-scripts/add-folder-roles-for-service-accounts-and-terra-billing-and-test-users.sh [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  echo "${VALID_ENVS}"
+  exit 1
+else
+  # set FOLDER_ID and SA email based on the env
+  if [[ "${ENV}" == "dev" ]] ; then
+    FOLDER_ID=599966635789 # folder: test.firecloud.org/dev
+  elif [[ "${ENV}" == "qa" ]]; then
+    FOLDER_ID=950154083315 # folder: test.firecloud.org/tools (no QA folder)
+  elif [[ "${ENV}" == "perf" ]]; then
+    FOLDER_ID=1076814209841 # folder: test.firecloud.org/perf
```
Comment (yonghaoy) [inline scripts/project-setup-scripts/add-folder-roles-for-service-accounts-and-terra-billing-and-test-users.sh:31]: this one folder is not inteneded to used by CWB, it is the folder which buffer's own perf test is using. I don't think you need this. For Rawsl/Terra's perf test, you may use qualify.fc/org. 
And also, that folder is being deprecated

Comment (andy7i) [reply scripts/project-setup-scripts/add-folder-roles-for-service-accounts-and-terra-billing-and-test-users.sh:31]: thanks for the heads up about `perf` being used for buffer perf testing! I didn't realize that was its purpose.

I think that only FiaBs use quality.fc.org, and real test envs use test.fc.org, so we probably want the perf env to use a folder in test.fc.org. 

Since the perf folder won't be used for the buffer perf tests, would it be ok for the perf env to use this folder?

Comment (yonghaoy) [reply scripts/project-setup-scripts/add-folder-roles-for-service-accounts-and-terra-billing-and-test-users.sh:36]: I think you can take that folder. But you may need to create a sub-folder called `CommunityWorkbench`.

Comment (andy7i) [reply scripts/project-setup-scripts/add-folder-roles-for-service-accounts-and-terra-billing-and-test-users.sh:36]: cool, looks like the perf/CommunityWorkbench already exists! https://console.cloud.google.com/iam-admin/settings?folder=246142551922

**Review by gpolumbo-broad — APPROVED** [review]

File: scripts/project-setup-scripts/add-folder-roles-for-service-accounts-and-terra-billing-and-test-users.sh (hunk 10 above)
Comment (yonghaoy) [reply scripts/project-setup-scripts/add-folder-roles-for-service-accounts-and-terra-billing-and-test-users.sh:36]: Is that will be used for perf Rawls talk to perf RBS? I can add a perf RBS pool using that folder.
Also, I think DevOps is working on setting up quality.firecloud.org folders creation
https://broadworkbench.atlassian.net/browse/DDO-1089

Comment (andy7i) [reply scripts/project-setup-scripts/add-folder-roles-for-service-accounts-and-terra-billing-and-test-users.sh:36]: I think that would be great. The quality.firecloud.org folder is only used by broad-dsde-qa FIABs and not the perf environment so we should keep those separate. Once DDO-1089 is done I can add the prod and QA folders. There's a ticket to track this

```
# todo: to be created as part of https://broadworkbench.atlassian.net/browse/CA-1194 after completion of https://broadworkbench.atlassian.net/browse/DDO-1089
```
